### ISSUE TITLE
Problem with Graph.add_shared_node()

### ISSUE THREAD

**Issue by geffy** [fchollet/keras#1171]
Hi, I try to add shared layer in Graph, but facing this problem:

```
from keras.models import Graph
from keras.layers import containers

block = containers.Sequential()
block.add(Dense(8, input_dim=20, activation='relu'))
block.add(Dense(1))


graph = Graph()
graph.add_input(name='input1', input_shape=(2,))
graph.add_input(name='input2', input_shape=(2,))

graph.add_shared_node(block, name='blocks', inputs=['input1', 'input2'], merge_mode='sum')
graph.add_node(Activation('sigmoid'), name='sigm', input='blocks')

graph.add_output(name='output', input='sigm')
graph.compile('rmsprop', {'output':'binary_crossentropy'})



---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-18-501bebb9a4df> in <module>()
      8 graph.add_input(name='input2', input_shape=(2,))
      9 
---> 10 graph.add_shared_node(block, name='blocks', inputs=['input1', 'input2'], merge_mode='sum')
     11 graph.add_node(Activation('sigmoid'), name='sigm', input='blocks')
     12 

/usr/local/lib/python2.7/site-packages/Keras-0.3.0-py2.7.egg/keras/layers/containers.pyc in add_shared_node(self, layer, name, inputs, merge_mode, concat_axis, dot_axes, outputs, create_output)
    327                 raise Exception('Unknown identifier: ' + input)
    328         s = Siamese(layer, layers, merge_mode, concat_axis=concat_axis, dot_axes=dot_axes)
--> 329         s.set_name(name)
    330         self.namespace.add(name)
    331         self.nodes[name] = s

AttributeError: 'Siamese' object has no attribute 'set_name'
```

Is it my misunderstanding of using pattern or problem in code?

**Comment by fchollet** [comment]
@farizrahman4u can you add unit tests for the untested code in Siamese, Lambda and LambdaMerge?

**Comment by farizrahman4u** [comment]
Yes.Will be done by tomorrow.

**Comment by farizrahman4u** [comment]
#1196

**Comment by farizrahman4u** [comment]
I am working on unit tests for siamese layer.

**Comment by jeffzhengye** [comment]
@farizrahman4u There are several bugs in Siamese. 
1.   File "/home/jeffye/workspace/jeff/keras/keras/layers/containers.py", line 351, in add_shared_node
    sh.set_name(sh_name)
AttributeError: 'SiameseHead' object has no attribute 'set_name'
I guess this is due to api change. 

2.  you cannot add parameters in layers to a Siamese layer (see the codes below) because a) params in layers could be untrainable b) will be duplicated for updates. 
I guess you copy the boilerplate codes from Merge, but Merge layer will not occur in graph.nodes (see def add_node in containers.py).  

       for l in self.layers:
            params, regs, consts, updates = l.get_params()
            self.regularizers += regs
            self.updates += updates
            # params and constraints have the same size
            for p, c in zip(params, consts):
                if p not in self.params:
                    self.params.append(p)

If I am wrong, please correct me. Thanks.

**Comment by elbamos** [comment]
Just +1'ing that I'm seeing the same bug as @geffy

**Comment by farizrahman4u** [comment]
@elbamos Reinstall  keras

**Comment by jeffzhengye** [comment]
@fchollet @farizrahman4u Do you think the following is a bug? I tested on sharing the embedding layer, for which I cannot set it to trainable=False. 

* you cannot add parameters in layers to a Siamese layer (see the codes below) because a) params in layers could be untrainable b) will be duplicated for updates. 
I guess you copy the boilerplate codes from Merge, but Merge layer will not occur in graph.nodes (see def add_node in containers.py).

for l in self.layers:
params, regs, consts, updates = l.get_params()
self.regularizers += regs
self.updates += updates
# params and constraints have the same size
for p, c in zip(params, consts):
if p not in self.params:
self.params.append(p)

If I am wrong, please correct me. Thanks.

**Comment by elbamos** [comment]
@farizrahman4u  Tried reinstalling keras, no effect:

ld = Dense(128, activation = "softmax", 
          init = 'uniform', 
          W_constraint = unitnorm(), 
          W_regularizer = l2(0.001)
          )

ntm.add_shared_node(ld, name = "ld", inputs = ["d_pos", "d_neg"],  outputs = ["ld_pos", "ld_neg"], merge_mode=None)


---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-7-20fa2b4c09d5> in <module>()
----> 1 ntm.add_shared_node(ld, name = "ld", inputs = ["d_pos", "d_neg"],  outputs = ["ld_pos", "ld_neg"], merge_mode=None)

/Volumes/home500/anaconda/envs/amos27/lib/python2.7/site-packages/keras/layers/containers.pyc in add_shared_node(self, layer, name, inputs, merge_mode, concat_axis, dot_axes, outputs, create_output)
    324                 raise Exception('Unknown identifier: ' + input)
    325         s = Siamese(layer, layers, merge_mode, concat_axis=concat_axis, dot_axes=dot_axes)
--> 326         s.set_name(name)
    327         self.namespace.add(name)
    328         self.nodes[name] = s

AttributeError: 'Siamese' object has no attribute 'set_name'


***
My languages are really R and Scala -- I'm only moving to Python and Keras to experiment with some particular generative NN models.  So, if you guys tell me that the reason I'm seeing this is I'm doing something impossibly stupid with either keras or Python, I totally won't be offended.

**Comment by farizrahman4u** [comment]
@elbamos Go to [containers.py](https://github.com/fchollet/keras/blob/master/keras/layers/containers.py), you will see that there is no 'set_name' anywhere in that file. So I am sure that you are not running the latest Keras. Mabe you cloned Keras, but forgot to run setup.py. Or maybe you are doing pip install, which is not up to date. Try the following:

```
 git clone http://www.github.com/fchollet/keras.git
 cd  keras
 python setup.py install
 pip install --upgrade --no-deps git+git://github.com/Theano/Theano.git
```

@jeffzhengye I will look into it..Also, what about merge layer? Arent params of the layers added irrespective of if  they are trainable or not?
Whatever, I will check it out.

**Comment by elbamos** [comment]
@farizrahman4u Yup, it was pip install.  Thanks!

**Comment by jeffzhengye** [comment]
@farizrahman4u For Merge layer, It won't cause this kind of problem because it is not in graph.nodes so the params will not be collected for update. It doesn't matter though the codes seems redundant in Merge. 

In your case (Siamese), Siamese will be added in graph.nodes, instead of layer (in add_shared_node). In add_node, layer param will be added in graph.nodes. So Keras will collect params from Siamese, instead of layer param. I think this is the root of the problem. 

Basically, we loss the control of setting trainable in your case. When you want Siamese to be trainable=True, it's fine. Otherwise, it should be a bug.

**Comment by farizrahman4u** [comment]
@jeffzhengye So  the only problem is when we want siamese to be not trainable right?

**Comment by jeffzhengye** [comment]
exactly. you take settings from layer, but forget trainable.
Maybe also other settings, which I didn't test yet.

On Wed, Dec 9, 2015 at 12:37 PM, Fariz Rahman <notifications@github.com>
wrote:

> @jeffzhengye <https://github.com/jeffzhengye> So the only problem is when
> we want siamese to be not trainable right?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/fchollet/keras/issues/1171#issuecomment-163334831>.
>



-- 
Jeff Ye.

**Comment by phreeza** [comment]
@jeffzhengye Any chance you could write a test to check for this behaviour? Seems like an important failure mode to catch...

**Comment by farizrahman4u** [comment]
I have a doubt. Assuming this is how weights are obtained for updating during training in a `Sequential` model:

```
weights = []
for every layer in model:
    if layer is trainable:
        weights += layer.weights
set update equations for every w in weights.
```

In a 2 legged sequential model, with one of the legs having un-trainable layers, wont the weights of the frozen layers still get  an update equation and be trained?

```python
s1=Sequential()
s2=Sequential()

s1.add(Layer(...))
s2.add(Layer(..., trainable=False))

m=Sequential()
m.add(Merge([s1,s2],'sum'))
m.add(Layer(...))
m.compile(...)
m.fit([x1,x2],y)
```

**Comment by elbamos** [comment]
I am still having trouble, this time with the "shared variable already has an update expression" error:

```python
ntm = Graph()
ntm.add_input(name='g', input_shape=(300,))  
ntm.add_input("d_pos", input_shape = (n_docs,))
ntm.add_input("d_neg", input_shape = (n_docs,))
ntm.add_node(Dense(n_topics, activation = "sigmoid", 
                   W_regularizer = l2(0.001)), 
             name = "lt", input = "g")
ntm.add_shared_node(Dense(128, activation = "softmax", 
                      W_regularizer = l2(0.001)
                      ), name = "ld", 
                    inputs = ["d_pos", "d_neg"],  
                    outputs = ["ld_pos", "ld_neg"], 
                    merge_mode=None)
ntm.add_shared_node(Flatten(), 
                    name = "ls_pos", 
                    inputs = ["lt", "ld_pos" ],
                    merge_mode = 'dot', dot_axes = 1)  
ntm.add_shared_node(Flatten(),
                    name = "ls_neg", 
                    inputs = ["lt", "ld_neg"], 
                    merge_mode = 'dot', dot_axes = 1) 
ntm.add_shared_node(ActivityRegularization(0,0), # A passthrough
                    name = "summed", 
                    inputs = ["ls_pos", "ls_neg"], 
                    merge_mode = 'sum') # output = (,1)
ntm.add_node(ThresholdedReLU(0.5), 
             name = "lossthreshold",
             input = "summed") # output = (,1)
ntm.add_output(name = "loss_out", 
               input= "lossthreshold")
import keras.backend as K
def obj(X, Y):
    return K.sum(Y)
ntm.compile(loss = {'loss_out' : obj}, optimizer = 'sgd')

ValueError: ('this shared variable already has an update expression', (<CudaNdarrayType(float32, matrix)>, GpuFromHost.0))
```

Before anyone asks -- I know it looks weird, its not my algorithm, I'm trying to implement this:  http://nlp.cs.rpi.edu/paper/AAAI15.pdf

**Comment by farizrahman4u** [comment]
@elbamos This error will be fixed soon. In the mean time, you can implement your model like this:

```python
from keras.models import*
from keras.layers.core import*
from keras.layers.embeddings import*
from keras.regularizers import*
from  keras.layers.advanced_activations import*
n_docs = 5
n_topics = 3
ntm = Graph()
ntm.add_input(name='g', input_shape=(300,))  
ntm.add_input("d_pos", input_shape = (n_docs,))
ntm.add_input("d_neg", input_shape = (n_docs,))
ntm.add_node(Dense(n_topics, activation = "sigmoid", 
                   W_regularizer = l2(0.001)), 
             name = "lt", input = "g")
ntm.add_shared_node(Dense(128, activation = "softmax", 
                      W_regularizer = l2(0.001)
                      ), name = "ld", 
                    inputs = ["d_pos", "d_neg"],  
                    outputs = ["ld_pos", "ld_neg"], 
                    merge_mode=None)
ntm.add_shared_node(Flatten(), 
                    name = "_ls_pos", 
                    inputs = ["lt", "ld_pos" ], outputs=["_lt", "_ld_pos"])
ntm.add_node(Layer(),name="ls_pos", inputs = ["_lt","_ld_pos"], merge_mode='dot', dot_axes=1)
ntm.add_shared_node(Flatten(),
                    name = "_ls_neg", 
                    inputs = ["lt", "ld_neg"], outputs=["_lt","_ld_neg"])
ntm.add_node(Layer(), name="ls_neg", inputs=["_lt","_ld_neg"],
	merge_mode='dot',dot_axes=1)
ntm.add_shared_node(ActivityRegularization(0,0), # A passthrough
                    name = "_summed", 
                    inputs = ["ls_pos", "ls_neg"], 
                    outputs = ["_ls_pos", "_ls_neg"]) # output = (,1)
ntm.add_node(Layer(), name="summed", inputs=["_ls_pos", "_ls_neg"],
	merge_mode='sum')
ntm.add_node(ThresholdedReLU(0.5), 
             name = "lossthreshold",
             input = "summed") # output = (,1)
ntm.add_output(name = "loss_out", 
               input= "lossthreshold")
import keras.backend as K
def obj(X, Y):
    return K.sum(Y)
ntm.compile(loss = {'loss_out' : obj}, optimizer = 'sgd')
```

Note  that there are still errors in your code which is not related to the `Siamese` layer.

**Comment by elbamos** [comment]
@farizrahman4u Thanks; could I ask you though to clarify the approach to working around the issue?

I think the intended structure of the original code may have been unclear, and that's why the proposed modification produces other errors, but if I understood the intent of the modification I might be able to get the whole thing to work. 

In the original code, there are a number of shared nodes whose only function is to take the dot-product of the output of two layers.  (The only node that's really "shared" in the sense of using shared weights is ld.  The concept is that ld should have two outputs.  Then each of those outputs is dot-multiplied by lt, producing a pair of scalars.)  

It looks like the intent of your fix is to replace those shared `Flatten` or `ActivityRegularization` nodes (which were just pass-throughs to the merge) with non-shared `Layer` nodes. 

Using the edit verbatim, of course I get an error that _lt is defined twice.  Modifying it to replace shared Flatten nodes with non-shared Layer nodes, which I think is your intent, and it seems to get a lot closer to compiling:

```
ntm = Graph()
ntm.add_input(name='g', input_shape=(300,))  
ntm.add_input("d_pos", input_shape = (n_docs,)) 
ntm.add_input("d_neg", input_shape = (n_docs,)) 
ntm.add_node(Dense(n_topics, activation = "sigmoid", 
             name = "lt", input = "g")
ntm.add_shared_node(Dense(128, activation = "softmax", 
                      ), name = "ld", 
                    inputs = ["d_pos", "d_neg"],  
                    outputs = ["ld_pos", "ld_neg"], 
                    merge_mode=None)
ntm.add_node(Flatten(), 
                    name = "_ld_pos", 
                    input = "ld_pos")
ntm.add_node(Layer(),
                   name = "ls_pos", 
                   inputs = ["lt", "_ld_pos"], 
                   merge_mode = 'dot', dot_axes = 1)
ntm.add_node(Flatten(),
                    name = "_ld_neg", 
                    input = "ld_neg") 
ntm.add_node(Layer(), 
                   name = "ls_neg", 
                   inputs = ["lt", "_ld_neg"], 
                    merge_mode = 'dot', dot_axes = 1) 
ntm.add_node(Layer(), name = "summed",
             inputs = ["ls_pos", "ls_neg"],
                    merge_mode = 'sum') 
ntm.add_node(ThresholdedReLU(0.5), 
             name = "lossthreshold",
             input = "summed") # output = (,1)
ntm.add_output(name = "loss_out", 
               input= "lossthreshold")

UnusedInputError: theano.function was asked to create a function computing outputs given certain inputs, but the provided input variable at index 3 is not part of the computational graph needed to compute the outputs: <TensorType(float32, matrix)>.
To make this error into a warning, you can pass the parameter on_unused_input='warn' to theano.function. To disable it completely, use on_unused_input='ignore'.
```

From the visualization of the graph, the only thing missing are the two inputs d_pos and d_neg.  But that part of the design seems pretty straightforward, and you didn't suggest changing anything there?

**Comment by farizrahman4u** [comment]
@elbamos The  only thing I did is that instead merging at the shared node, I merge them in a separate non shared layer to bypass an error.

**Comment by farizrahman4u** [comment]
Ok, everything will be fixed with  #1271.

**Comment by elbamos** [comment]
Yes, all fixed, thanks!

**Comment by farizrahman4u** [comment]
@elbamos Cool!

Do you  have anything good that could serve as an example of Siamese layer? Anything that works on the datasets already available in Keras?

**Comment by elbamos** [comment]
@farizrahman4u Well, I'm continuing to actively work on this: https://github.com/elbamos/NeuralTopicModels

I could play with it and try to get it to work on one of the built-in datasets.  Actually that's not a bad idea at all because the IMDB review dataset is one of the test sets that was used in the paper, and it would give me an excuse to learn keras' text pre-processing tools.  

I'll try to put something together when my GPU becomes available.

**Comment by cxfneo** [comment]
@farizrahman4u @fchollet 
I am still find a similar error or bug as above and I update my library to latest one. I am very confused now
I build a graph which will be as a shared node of another graph with multi-input. Then it gave error info as follows:
"theano.compile.function_module.UnusedInputError: theano.function was asked to create a function computing outputs given certain inputs, but the provided input variable at index 1 is not part of the computational graph needed to compute the outputs: input_b.
To make this error into a warning, you can pass the parameter on_unused_input='warn' to theano.function. To disable it completely, use on_unused_input='ignore'."
------------------------------------------------------------------------------------------------------------------------ 
The code is below: I want to implement an triple-loss. The base network is a LSTM graph
def create_base_network(last_hidden_dims,hidden_units,vocabsize,timesteps,data_dim):
    model = Graph()
    model.add_input(name='input0', input_shape=(timesteps,), dtype=int)
    model.add_node(Embedding(vocabsize, data_dim, input_length=timesteps),
               name='embedding', input='input0')
    model.add_node(LSTM(hidden_units), name='forward', input='embedding')
    model.add_node(Dense(last_hidden_dims, activation='sigmoid'), name='sigmoid', input='forward')
    model.add_output(name='output', input='sigmoid')
    return model
--------------------------------------------------------------------------------------------------------------
base_network = create_base_network(129,256,vocabsize,nsteps,datadim)      
g = Graph()
g.add_input(name='input_a', input_shape=(nsteps,),dtype='int')
g.add_input(name='input_b', input_shape=(nsteps,),dtype='int')
g.add_input(name='input_c', input_shape=(nsteps,),dtype='int')
g.add_shared_node(base_network, name='shared', inputs=['input_a','input_b','input_c'],merge_mode='join')
-------------------------------------------------------------------------------------------------------------
g.add_node(Lambda(euclidean_distance),name='d',inputs='shared')
g.add_output(name='output', input='d')
rms = RMSprop()
g.compile(loss={'output': contrastive_loss}, optimizer=rms)
-----------------------------------------------------------------------------------------------------------------
def euclidean_distance(inputs):
    print (type(inputs))
    assert len(inputs) == 3, ('Euclidean distance needs '
                              '3 inputs, %d given' % len(inputs))
    u, v, w = inputs.values()
    u=K.l2_normalize(u, axis=1)
    v=K.l2_normalize(v, axis=1)
    w=K.l2_normalize(w, axis=1)
    cos1=u*v
    cos2=u*w
    return K.concatenate([cos1,cos2],axis=0) // how to return with two values?
def contrastive_loss(y, d):
    size=y.shape[0]
    s1=d[0:size]
    s2=d[size:]
    diff=(s1-s2)*(-20.0)
    return diff[0]+y[0]*0.0
--------------------------------------------------------------------------------------------------------
